{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import glob\n",
    "import numpy\n",
    "import scipy\n",
    "import h5py\n",
    "\n",
    "from scipy import sparse\n",
    "\n",
    "from dragonnfruit.io import LocusGenerator\n",
    "from dragonnfruit.io import GenomewideGenerator\n",
    "\n",
    "from dragonnfruit.models import CellStateController\n",
    "from dragonnfruit.models import DynamicBPNet\n",
    "from dragonnfruit.models import DragoNNFruit\n",
    "from dragonnfruit.preprocessing import extract_fragments\n",
    "from dragonnfruit.preprocessing import preprocess_sparse_atac\n",
    "from dragonnfruit.preprocessing import read_chrom_sizes\n",
    "from dragonnfruit.io import save_data, load_data\n",
    "\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import muon as mu\n",
    "from muon import atac as ac\n",
    "\n",
    "import collections\n",
    "from scipy.sparse import csc_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "44110005it [02:07, 345695.72it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "/workspaces/torch_ddsm/_data_pool1/10x_data/pbmc3kv2/pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz\n",
      "Cell included:  461847\n",
      "Cells excluded:  0\n",
      "Cells not included:  0\n",
      "Cells added:  461847\n",
      "Cell inclusion list size:  0\n",
      "Cell exclusion list size:  0\n",
      "Frag. included count:  43986816\n",
      "Frag. filtered by chrom:  16397\n",
      "Frag. filtered by length:  106741\n",
      "\n"
     ]
    }
   ],
   "source": [
    "chrom_sizes, header = read_chrom_sizes(\"/workspaces/torch_ddsm/_data_pool1/general/hg38.chrom.sizes\")\n",
    "canonical_chroms = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', \n",
    "\t\t'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', \n",
    "\t\t'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', \n",
    "\t\t'chr22', 'chrX']\n",
    "\n",
    "filtered_chrom_sizes = {key: value for key, value in chrom_sizes.items() if key in canonical_chroms}\n",
    "\n",
    "chroms = list(filtered_chrom_sizes.keys())\n",
    "\n",
    "parameters = {\n",
    "\t'fragments': [\"/workspaces/torch_ddsm/_data_pool1/10x_data/pbmc3kv2/pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz\"],\n",
    "\t'sequences': \"/workspaces/torch_ddsm/_data_pool1/cellranger_references/refdata-gex-GRCh38-2020-A/fasta/genome.fa\",\n",
    "\t'chrom_sizes': filtered_chrom_sizes,\n",
    "\t'chroms': chroms,\n",
    "\t'include_cells': None,\n",
    "\t'exclude_cells': None,\n",
    "\t'cell_name_prefixes': '',\n",
    "\t'max_fragment_length': 1000,\n",
    "\t'start_offset': 4,\n",
    "\t'end_offset': -5,\n",
    "\t'n_jobs': 16,\n",
    "\t'verbose': True,\n",
    "\t'signals':'dragonnfruit_data.h5',\n",
    "\t'read_depths':'atac_read_depths.npz',\n",
    "\t'save_dir': '/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc',\n",
    "    'loci': '/workspaces/torch_ddsm/_data_pool1/10x_data/pbmc3kv2/pbmc_granulocyte_sorted_3k_atac_peaks.bed',\n",
    "    'n_neighbors': 500,\n",
    "\t'n_components': 50,\n",
    " \t'neighbors': 'atac_neighbors.npz',\n",
    "  \t'pca': 'atac_pca.npz',\n",
    "\t'count_matrix': 'atac_cellxpeak_counts.npz',\n",
    "\t'name' : \"pbmc_granulocyte_sorted_3k_atac\",\n",
    "}\n",
    "X_cscs, read_depths = extract_fragments(\n",
    "\t\tfragments=parameters['fragments'],\n",
    "\t\tchrom_sizes=parameters['chrom_sizes'],\n",
    "\t\tchroms=parameters['chroms'],\n",
    "\t\tinclude_cells=parameters['include_cells'],\n",
    "\t\texclude_cells=parameters['exclude_cells'],\n",
    "\t\tcell_name_prefixes=parameters['cell_name_prefixes'],\n",
    "\t\tmax_fragment_length=parameters['max_fragment_length'],\n",
    "\t\tstart_offset=parameters['start_offset'],\n",
    "\t\tend_offset=parameters['end_offset'],\n",
    "\t\tn_jobs=parameters['n_jobs'],\n",
    "\t\tverbose=parameters['verbose']\n",
    ")\n",
    "\n",
    "os.makedirs(parameters['save_dir'], exist_ok=True)\n",
    "save_data(f\"{parameters['save_dir']}/{parameters['signals']}\", X_cscs)\n",
    "numpy.savez_compressed(f\"{parameters['save_dir']}/{parameters['read_depths']}\", read_depths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "98319it [00:08, 11377.58it/s]\n"
     ]
    }
   ],
   "source": [
    "cellxlocus, X_pca, neighbors = preprocess_sparse_atac(\n",
    "    X_cscs, \n",
    "    peaks=parameters['loci'], \n",
    "    chroms=parameters['chroms'],\n",
    "    n_components=parameters['n_components'], \n",
    "    n_neighbors=parameters['n_neighbors'],\n",
    "    verbose=parameters['verbose'])\n",
    "\n",
    "scipy.sparse.save_npz(f\"{parameters['save_dir']}/{parameters['count_matrix']}\", cellxlocus)\n",
    "numpy.savez_compressed(f\"{parameters['save_dir']}/{parameters['pca']}\", X_pca)\n",
    "numpy.savez_compressed(f\"{parameters['save_dir']}/{parameters['neighbors']}\", neighbors)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chrom pseudobulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dragonnfruit.preprocessing import create_pseudobulks\n",
    "create_pseudobulks(\n",
    "    reads=X_cscs, \n",
    "    output_filename=f\"{parameters['save_dir']}/pbulk.bw\", \n",
    "    chrom_sizes=parameters['chrom_sizes'],\n",
    "    chroms=parameters['chroms'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<461847x248956422 sparse matrix of type '<class 'numpy.int8'>'\n",
       "\twith 8245994 stored elements in Compressed Sparse Column format>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_cscs['chr1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fit bias model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Does this need to be cell-wise bias?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Loci:   0%|          | 166/98319 [00:00<00:59, 1654.05it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Loci: 100%|██████████| 98319/98319 [01:04<00:00, 1522.82it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: chrY 11295191 11296191 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: chrY 11332699 11333699 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: chrY 11333520 11334520 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: chrY 56836164 56837164 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270727.1 52033 53033 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000194.1 27836 28836 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000194.1 100863 101863 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000194.1 114440 115440 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000205.2 62884 63884 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000205.2 67693 68693 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000205.2 68930 69930 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000205.2 87394 88394 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000205.2 88575 89575 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000205.2 140285 141285 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000195.1 30343 31343 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000195.1 32118 33118 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 39888 40888 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 42107 43107 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 44581 45581 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 45641 46641 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 50677 51677 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 90012 91012 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 92113 93113 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 99210 100210 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: GL000219.1 124953 125953 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270734.1 133437 134437 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270721.1 2034 3034 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270726.1 27095 28095 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270726.1 41409 42409 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270713.1 3901 4901 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270713.1 21386 22386 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270713.1 29489 30489 not valid bigwig indexes. Using zeros instead.\n",
      "Warning: KI270713.1 36874 37874 not valid bigwig indexes. Using zeros instead.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 33/33 [00:00<00:00, 16678.16it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC Bin\tBackground Count\tPeak Count\tChosen Count\n",
      "0.00:        0\t       0\t       0\n",
      "0.02:        0\t       0\t       0\n",
      "0.04:        0\t       0\t       0\n",
      "0.06:        6\t       0\t       0\n",
      "0.08:       10\t       0\t       0\n",
      "0.10:       12\t       0\t       0\n",
      "0.12:       17\t       0\t       0\n",
      "0.14:       27\t       0\t       0\n",
      "0.16:       39\t       0\t       0\n",
      "0.18:       36\t       0\t       0\n",
      "0.20:       70\t       0\t       0\n",
      "0.22:      124\t       4\t       4\n",
      "0.24:      236\t       7\t       7\n",
      "0.26:     1362\t      12\t      12\n",
      "0.28:     8227\t      56\t      56\n",
      "0.30:    30605\t     275\t     275\n",
      "0.32:    70229\t    1004\t    1004\n",
      "0.34:   116803\t    2629\t    2629\n",
      "0.36:   157100\t    4986\t    4986\n",
      "0.38:   179921\t    7274\t    7274\n",
      "0.40:   153826\t    9064\t    9064\n",
      "0.42:   111868\t   10147\t   10147\n",
      "0.44:    83789\t    9823\t    9823\n",
      "0.46:    60757\t    9025\t    9025\n",
      "0.48:    38721\t    7931\t    7931\n",
      "0.50:    23670\t    7149\t    7149\n",
      "0.52:    14046\t    5870\t    6010\n",
      "0.54:     8599\t    5069\t    8599\n",
      "0.56:     5702\t    4221\t    5702\n",
      "0.58:     3658\t    3706\t    3658\n",
      "0.60:     2135\t    2955\t    2135\n",
      "0.62:     1290\t    2358\t    1290\n",
      "0.64:      738\t    1813\t     738\n",
      "0.66:      370\t    1258\t     370\n",
      "0.68:      223\t     787\t     223\n",
      "0.70:       83\t     476\t      83\n",
      "0.72:       68\t     258\t      68\n",
      "0.74:       32\t     101\t      32\n",
      "0.76:       13\t      45\t      13\n",
      "0.78:        5\t      13\t       5\n",
      "0.80:        6\t       2\t       6\n",
      "0.82:        0\t       1\t       0\n",
      "0.84:        1\t       0\t       1\n",
      "0.86:        0\t       0\t       0\n",
      "0.88:        0\t       0\t       0\n",
      "0.90:        0\t       0\t       0\n",
      "0.92:        0\t       0\t       0\n",
      "0.94:        0\t       0\t       0\n",
      "0.96:        0\t       0\t       0\n",
      "0.98:        0\t       0\t       0\n",
      "1.00:        0\t       0\t       0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Loci: 100%|██████████| 98319/98319 [01:27<00:00, 1128.61it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC-bin KS test stat:0.0524, p-value 1.03e-117\n",
      "Peak Robust Signal Minimum: 33.0\n",
      "Matched Signal Maximum: 16.0\n"
     ]
    }
   ],
   "source": [
    "from bpnetlite.negatives import extract_matching_loci\n",
    "\n",
    "bp_parameters = {\n",
    "\t'peaks': '/workspaces/torch_ddsm/_data_pool1/10x_data/pbmc3kv2/pbmc_granulocyte_sorted_3k_atac_peaks.bed',\n",
    "\t'fasta': parameters['sequences'],\n",
    "\t'bin_width': 0.02,\n",
    "\t'max_n_perc': 0.1,\n",
    "\t'bigwig': '/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbulk.bw',\n",
    "\t'beta': 0.5,\n",
    "\t'in_window': 2114,\n",
    "\t'out_window': 1000,\n",
    "\t'verbose': True,\n",
    "    'chroms': chroms,\n",
    "    'max_jitter': 128,\n",
    "    'output': 'matched_loci.bed',\n",
    "}\n",
    "\n",
    "# Extract regions that match the GC content of the peaks\n",
    "matched_loci = extract_matching_loci(\n",
    "    loci=parameters['loci'], \n",
    "    fasta=bp_parameters['fasta'],\n",
    "    gc_bin_width=bp_parameters['bin_width'],\n",
    "    max_n_perc=bp_parameters['max_n_perc'],\n",
    "    bigwig=bp_parameters['bigwig'],\n",
    "    signal_beta=bp_parameters['beta'],\n",
    "    in_window=bp_parameters['in_window'],\n",
    "    out_window=bp_parameters['out_window'],\n",
    "    verbose=bp_parameters['verbose']\n",
    ")\n",
    "\n",
    "matched_loci.to_csv(f\"{parameters['save_dir']}/{bp_parameters['output']}\", header=False, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Train the bias model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Loci:   0%|          | 0/82087 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Loci: 100%|██████████| 82087/82087 [01:02<00:00, 1306.31it/s]\n",
      "Loading Loci: 100%|██████████| 16232/16232 [00:12<00:00, 1315.10it/s]\n"
     ]
    }
   ],
   "source": [
    "# default bias parameters\n",
    "from bpnetlite.io import PeakGenerator\n",
    "from bpnetlite.io import extract_loci\n",
    "\n",
    "\n",
    "training_data = PeakGenerator(\n",
    "    loci=matched_loci, \n",
    "    sequences=parameters['sequences'],\n",
    "    signals=[f\"{parameters['save_dir']}/pbulk.bw\"],\n",
    "    controls=None,\n",
    "    chroms=chroms[2:],\n",
    "    in_window=bp_parameters['in_window'],\n",
    "    out_window=bp_parameters['out_window'],\n",
    "    max_jitter=bp_parameters['max_jitter'],\n",
    "    reverse_complement=True,\n",
    "    min_counts=None,\n",
    "    max_counts=None,\n",
    "    random_state=None,\n",
    "    batch_size=64,\n",
    "    verbose=True\n",
    ")\n",
    "\n",
    "valid_data = extract_loci(\n",
    "    sequences=parameters['sequences'],\n",
    "    signals=[f\"{parameters['save_dir']}/pbulk.bw\"],\n",
    "    controls=None,\n",
    "    loci=matched_loci,\n",
    "    chroms=chroms[:2],\n",
    "    in_window=bp_parameters['in_window'],\n",
    "    out_window=bp_parameters['out_window'],\n",
    "    max_jitter=bp_parameters['max_jitter'],\n",
    "    verbose=True,\n",
    ")\n",
    "\n",
    "valid_sequences, valid_signals = valid_data\n",
    "valid_controls = None\n",
    "n_control_tracks = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "bias_fit_parameters = {\n",
    "    \"n_filters\": 256,\n",
    "    \"n_layers\": 4,\n",
    "    \"max_counts\": None,\n",
    "    \"loci\": \"../../../chromatin-atlas/ATAC/ENCSR637XSC/negatives_data/negatives.bed\",\n",
    "    \"verbose\": True,\n",
    "    \"random_state\": None,\n",
    "\t\"alpha\": 10,\n",
    "\t\"beta\": 0.5,\n",
    "    \"lr\": 0.000001,\n",
    "    \"max_epochs\": 10,\n",
    "    \"batch_size\": 512,\n",
    "    \"n_control_tracks\": n_control_tracks,\n",
    "    \"valid_controls\": valid_controls,\n",
    "    \"n_outputs\" : 1,\n",
    "    \"validation_iter\": 1000,\n",
    "}\n",
    "\n",
    "# bias_fit_parameters = parameters.copy()\n",
    "\n",
    "name = f\"{parameters['name']}.chrombpnet.bias.fit.json\"\n",
    "bias_fit_parameters['name'] = f\"{parameters['name']}.bias\"\n",
    "parameters['bias_model'] = f\"{bias_fit_parameters['name']}.torch\"\n",
    "\n",
    "min_counts = training_data.dataset.signals.sum(dim=(1, 2)).min().item()\n",
    "bias_fit_parameters['max_counts'] = min_counts * bias_fit_parameters['beta']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run BPnet fit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch\tIteration\tTraining Time\tValidation Time\tTraining MNLL\tTraining Count MSE\tValidation MNLL\tValidation Profile Pearson\tValidation Count Pearson\tValidation Count MSE\tSaved?\n",
      "0\t0\t1.299\t3.2782\t37.5275\t4.6096\t45.9182\t0.0011741763\t0.12038599\t5.3281\tTrue\n",
      "0\t1000\t36.343\t2.7568\t33.5983\t0.5789\t44.6027\t0.076191746\t-0.091623284\t0.8147\tTrue\n",
      "Epoch 0: Average Validation Loss = 88.95395052367974\n",
      "1\t2000\t23.7664\t2.7651\t35.3002\t0.6034\t43.1947\t0.097735494\t0.11622072\t0.8001\tTrue\n",
      "Epoch 1: Average Validation Loss = 52.064289390343944\n",
      "2\t3000\t14.4085\t2.7706\t33.318\t0.7486\t41.7538\t0.107549556\t0.1784533\t0.7775\tTrue\n",
      "Epoch 2: Average Validation Loss = 50.09292374192414\n",
      "3\t4000\t5.0294\t2.7607\t33.2003\t0.8144\t40.3799\t0.11340245\t0.18840684\t0.783\tTrue\n",
      "3\t5000\t35.8815\t2.7702\t27.0193\t0.7897\t39.4532\t0.11583789\t0.1907709\t0.7764\tTrue\n",
      "Epoch 3: Average Validation Loss = 48.262766575683216\n",
      "4\t6000\t28.7576\t2.7705\t31.039\t0.7894\t38.9984\t0.116843775\t0.19193283\t0.766\tTrue\n",
      "Epoch 4: Average Validation Loss = 47.03655629741523\n",
      "5\t7000\t19.4038\t2.7673\t32.7411\t0.4326\t38.7955\t0.11748404\t0.19266693\t0.7781\tTrue\n",
      "Epoch 5: Average Validation Loss = 46.61402943909215\n",
      "6\t8000\t10.0355\t2.7722\t28.8591\t0.7636\t38.702\t0.11796901\t0.19316925\t0.7754\tTrue\n",
      "Epoch 6: Average Validation Loss = 46.4842335023828\n",
      "7\t9000\t0.6538\t2.7618\t34.4002\t0.8374\t38.6542\t0.118295304\t0.19368361\t0.7816\tFalse\n",
      "7\t10000\t35.8622\t2.7637\t31.5372\t0.5329\t38.6252\t0.11851925\t0.19423538\t0.768\tTrue\n",
      "Epoch 7: Average Validation Loss = 46.436209520323615\n",
      "8\t11000\t24.3892\t2.7635\t30.4622\t0.6909\t38.6034\t0.118812695\t0.19467129\t0.7661\tTrue\n",
      "Epoch 8: Average Validation Loss = 46.28790004608321\n",
      "9\t12000\t15.0335\t2.7646\t26.5056\t0.7321\t38.5855\t0.11908316\t0.19510984\t0.7797\tFalse\n",
      "Epoch 9: Average Validation Loss = 46.34070309752705\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "from bpnetlite.bpnet import BPNet\n",
    "\n",
    "trimming = (bp_parameters['in_window'] - bp_parameters['out_window']) // 2\n",
    "\n",
    "model = BPNet(n_filters=bias_fit_parameters['n_filters'], \n",
    "    n_layers=bias_fit_parameters['n_layers'],\n",
    "    profile_output_bias=True,\n",
    "    count_output_bias=True,\n",
    "    alpha=bias_fit_parameters['alpha'],\n",
    "    trimming=trimming,\n",
    "    name=f\"{parameters['save_dir']}/{parameters['name']}\",\n",
    "    n_control_tracks=bias_fit_parameters['n_control_tracks'],\n",
    "    n_outputs=bias_fit_parameters['n_outputs']).cuda()\n",
    "\n",
    "optimizer = torch.optim.AdamW(model.parameters(), lr=bias_fit_parameters['lr'])\n",
    "\n",
    "model.fit(training_data, optimizer, X_valid=valid_sequences, \n",
    "\t\tX_ctl_valid=valid_controls, y_valid=valid_signals, max_epochs=bias_fit_parameters['max_epochs'], \n",
    "    batch_size=bias_fit_parameters['batch_size'],validation_iter=bias_fit_parameters['validation_iter'],verbose=bias_fit_parameters['verbose'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run attribution/motif diagnosis on the bias model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Attribution params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_parameters = {\n",
    "\t'sequences': parameters['sequences'],\n",
    "\t'loci': parameters['loci'],\n",
    " \t'batch_size': 5,\n",
    "\t'in_window': 2114,\n",
    "\t'out_window': 1000,\n",
    "\t'verbose': True,\n",
    "\t'max_jitter': 0,\n",
    "\t'chroms': ['chr6', 'chr7', 'chrX'],\n",
    "\t'model': None,\n",
    "\t'output': 'profile',\n",
    "\t'n_shuffles':20,\n",
    "\t'random_state':0,\n",
    " \t'ohe_filename': f\"{parameters['save_dir']}/{parameters['name']}.ohe.npz\",\n",
    "\t'attr_filename': f\"{parameters['save_dir']}/{parameters['name']}.attr.npz\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run attribution on the bias model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_parameters['sequences']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Loci:   0%|          | 0/13711 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Loci: 100%|██████████| 13711/13711 [00:26<00:00, 517.29it/s]\n"
     ]
    }
   ],
   "source": [
    "from bpnetlite.attributions import attribute\n",
    "\n",
    "model = torch.load(f\"{parameters['save_dir']}/{parameters['name']}.final.torch\").cuda()\n",
    "\n",
    "X = extract_loci(\n",
    "    sequences=attribute_parameters['sequences'],\n",
    "    loci=attribute_parameters['loci'],\n",
    "    chroms=attribute_parameters['chroms'],\n",
    "    max_jitter=attribute_parameters['max_jitter'],\n",
    "    verbose=attribute_parameters['verbose']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2743/2743 [05:54<00:00,  7.73it/s]\n"
     ]
    }
   ],
   "source": [
    "X_ctl = None\n",
    "X_attr = attribute(model, X, args=X_ctl,\n",
    "    model_output=attribute_parameters['output'], hypothetical=True, \n",
    "    n_shuffles=attribute_parameters['n_shuffles'],\n",
    "    batch_size=attribute_parameters['batch_size'],\n",
    "    random_state=attribute_parameters['random_state'],\n",
    "    verbose=attribute_parameters['verbose'])\n",
    "\n",
    "numpy.savez_compressed(attribute_parameters['ohe_filename'], X)\n",
    "numpy.savez_compressed(attribute_parameters['attr_filename'], X_attr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Modisco params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "modisco_motifs_parameters = {\n",
    "    'n_seqlets': 100000,\n",
    "    'output_filename': f\"{parameters['save_dir']}/{parameters['name']}_modisco_results.h5\",\n",
    "    'verbose': True,\n",
    "    'n_leiden': 2,\n",
    "    'window': 400,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run modisco on the attributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using 11910 positive seqlets\n",
      "Extracted 2033 negative seqlets\n"
     ]
    }
   ],
   "source": [
    "import modiscolite\n",
    "from modiscolite.util import calculate_window_offsets\n",
    "\n",
    "sequences = np.load(attribute_parameters['ohe_filename'])['arr_0']\n",
    "attributions = np.load(attribute_parameters['attr_filename'])['arr_0']\n",
    "\n",
    "center = sequences.shape[2] // 2\n",
    "start, end = calculate_window_offsets(center, modisco_motifs_parameters['window'])\n",
    "\n",
    "sequences = sequences[:, :, start:end].transpose(0, 2, 1)\n",
    "attributions = attributions[:, :, start:end].transpose(0, 2, 1)\n",
    "\n",
    "if sequences.shape[1] < modisco_motifs_parameters['window']:\n",
    "    raise ValueError(\"Window ({}) cannot be \".format(modisco_motifs_parameters['window']) +\n",
    "        \"longer than the sequences\".format(sequences.shape))\n",
    "\n",
    "sequences = sequences.astype('float32')\n",
    "attributions = attributions.astype('float32')\n",
    "\n",
    "pos_patterns, neg_patterns = modiscolite.tfmodisco.TFMoDISco(\n",
    "    hypothetical_contribs=attributions, \n",
    "    one_hot=sequences,\n",
    "    max_seqlets_per_metacluster=modisco_motifs_parameters['n_seqlets'],\n",
    "    sliding_window_size=20,\n",
    "    flank_size=5,\n",
    "    target_seqlet_fdr=0.05,\n",
    "    n_leiden_runs=modisco_motifs_parameters['n_leiden'],\n",
    "    verbose=modisco_motifs_parameters['verbose'])\n",
    "\n",
    "modiscolite.io.save_hdf5(modisco_motifs_parameters['output_filename'], pos_patterns, neg_patterns, modisco_motifs_parameters['window'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parameterize the modisco report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco_results.h5'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "modisco_motifs_parameters['output_filename']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "modisco_report_parameters= {\n",
    "    'motifs': '/workspaces/torch_ddsm/_data_pool1/jaspar/JASPAR2020_CORE_vertebrates_non-redundant_pfms_meme.txt',\n",
    "    'output_folder': f\"{parameters['save_dir']}/{parameters['name']}_modisco/\",\n",
    "    'verbose': True,\n",
    "    'input': modisco_motifs_parameters['output_filename'],\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate the modisco report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 5: Generate the tf-modisco report\n",
    "modiscolite.report.report_motifs(\n",
    "    modisco_report_parameters['input'], \n",
    "    modisco_report_parameters['output_folder'],\n",
    "    img_path_suffix=modisco_report_parameters['output_folder'], \n",
    "    meme_motif_db=modisco_report_parameters['motifs'],\n",
    "    is_writing_tomtom_matrix=True,\n",
    "    top_n_matches=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>pattern</th>\n",
       "      <th>num_seqlets</th>\n",
       "      <th>modisco_cwm_fwd</th>\n",
       "      <th>modisco_cwm_rev</th>\n",
       "      <th>match0</th>\n",
       "      <th>qval0</th>\n",
       "      <th>match0_logo</th>\n",
       "      <th>match1</th>\n",
       "      <th>qval1</th>\n",
       "      <th>match1_logo</th>\n",
       "      <th>match2</th>\n",
       "      <th>qval2</th>\n",
       "      <th>match2_logo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_0</td>\n",
       "      <td>3832</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_0.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_0.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0146.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0146.2.png\" width=\"240\" ></td>\n",
       "      <td>MA1600.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1600.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1619.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1619.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_1</td>\n",
       "      <td>600</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_1.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_1.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0836.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0836.2.png\" width=\"240\" ></td>\n",
       "      <td>MA0102.4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0102.4.png\" width=\"240\" ></td>\n",
       "      <td>MA0800.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0800.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_2</td>\n",
       "      <td>583</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_2.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_2.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0003.4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0003.4.png\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_3</td>\n",
       "      <td>539</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_3.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_3.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0116.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0116.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0750.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0750.2.png\" width=\"240\" ></td>\n",
       "      <td>MA1619.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1619.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_4</td>\n",
       "      <td>499</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_4.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_4.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA1619.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1619.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1615.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1615.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0017.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0017.2.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_5</td>\n",
       "      <td>308</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_5.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_5.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0163.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0163.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0017.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0017.2.png\" width=\"240\" ></td>\n",
       "      <td>MA0116.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0116.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_6</td>\n",
       "      <td>307</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_6.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_6.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0163.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0163.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0146.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0146.2.png\" width=\"240\" ></td>\n",
       "      <td>MA0809.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0809.2.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_7</td>\n",
       "      <td>237</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_7.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_7.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0812.1</td>\n",
       "      <td>0.999999</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0812.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0003.4</td>\n",
       "      <td>0.999999</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0003.4.png\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_8</td>\n",
       "      <td>182</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_8.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_8.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0146.2</td>\n",
       "      <td>0.471498</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0146.2.png\" width=\"240\" ></td>\n",
       "      <td>MA1615.1</td>\n",
       "      <td>0.471498</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1615.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1643.1</td>\n",
       "      <td>0.471498</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1643.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_9</td>\n",
       "      <td>153</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_9.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_9.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0781.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0781.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0779.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0779.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0138.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0138.2.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_10</td>\n",
       "      <td>149</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_10.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_10.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0003.4</td>\n",
       "      <td>0.120462</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0003.4.png\" width=\"240\" ></td>\n",
       "      <td>MA0812.1</td>\n",
       "      <td>0.300952</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0812.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0814.2</td>\n",
       "      <td>0.982995</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0814.2.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_11</td>\n",
       "      <td>138</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_11.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_11.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0808.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0808.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0774.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0774.1.png\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_12</td>\n",
       "      <td>121</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_12.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_12.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA1155.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1155.1.png\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_13</td>\n",
       "      <td>87</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_13.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_13.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0116.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0116.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1567.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1567.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1566.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1566.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_14</td>\n",
       "      <td>75</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_14.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_14.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0017.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0017.2.png\" width=\"240\" ></td>\n",
       "      <td>MA0836.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0836.2.png\" width=\"240\" ></td>\n",
       "      <td>MA0838.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0838.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_15</td>\n",
       "      <td>72</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_15.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_15.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0017.2</td>\n",
       "      <td>0.476773</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0017.2.png\" width=\"240\" ></td>\n",
       "      <td>MA1111.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1111.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0115.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0115.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_16</td>\n",
       "      <td>58</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_16.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_16.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA1619.1</td>\n",
       "      <td>0.758119</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1619.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1641.1</td>\n",
       "      <td>0.820812</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1641.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0150.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0150.2.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_17</td>\n",
       "      <td>39</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_17.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_17.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0142.1</td>\n",
       "      <td>0.127191</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0142.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1112.2</td>\n",
       "      <td>0.857070</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1112.2.png\" width=\"240\" ></td>\n",
       "      <td>MA0151.1</td>\n",
       "      <td>0.857070</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0151.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_18</td>\n",
       "      <td>35</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_18.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_18.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0140.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0140.2.png\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pos_patterns.pattern_19</td>\n",
       "      <td>32</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_19.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/pos_patterns.pattern_19.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0138.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0138.2.png\" width=\"240\" ></td>\n",
       "      <td>MA0748.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0748.2.png\" width=\"240\" ></td>\n",
       "      <td>MA1584.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1584.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>neg_patterns.pattern_0</td>\n",
       "      <td>55</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_0.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_0.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA1647.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1647.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1655.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1655.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0795.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0795.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>neg_patterns.pattern_1</td>\n",
       "      <td>45</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_1.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_1.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA1600.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1600.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0154.4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0154.4.png\" width=\"240\" ></td>\n",
       "      <td>MA1597.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1597.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>neg_patterns.pattern_2</td>\n",
       "      <td>40</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_2.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_2.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0496.3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0496.3.png\" width=\"240\" ></td>\n",
       "      <td>MA0781.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0781.1.png\" width=\"240\" ></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td><img src=\"NA\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>neg_patterns.pattern_3</td>\n",
       "      <td>38</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_3.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_3.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0114.4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0114.4.png\" width=\"240\" ></td>\n",
       "      <td>MA0484.2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0484.2.png\" width=\"240\" ></td>\n",
       "      <td>MA1619.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1619.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>neg_patterns.pattern_4</td>\n",
       "      <td>30</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_4.cwm.fwd.png\" width=\"240\" ></td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/trimmed_logos/neg_patterns.pattern_4.cwm.rev.png\" width=\"240\" ></td>\n",
       "      <td>MA0677.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0677.1.png\" width=\"240\" ></td>\n",
       "      <td>MA0115.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA0115.1.png\" width=\"240\" ></td>\n",
       "      <td>MA1148.1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td><img src=\"/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/MA1148.1.png\" width=\"240\" ></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "\n",
    "# Open the HTML file and read its content\n",
    "with open(\n",
    "    '/workspaces/torch_ddsm/_data_pool1/test_dragonnfruit_pbmc/pbmc_granulocyte_sorted_3k_atac_modisco/motifs.html', \n",
    "    'r') as file:\n",
    "    html_content = file.read()\n",
    "\n",
    "# Display the HTML content\n",
    "HTML(html_content)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "myenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
